Support typing.Union/typing.Optional in isinstance check (/support old-style unions)

**Preamble**

- This issue may already exists but I was unable to find an exact match. There are plenty of issues related to the PEP 604 syntax but none that are really about using `typing.Union`/`typing.Optional`/...
- Support for PEP 604 syntax was recently added in  #11673 / #17371 
- Old style unions were briefly mentioned in https://github.com/python/mypy/pull/17371#pullrequestreview-2115445793

**Feature**

Example code

```python
from typing import Any, Union, Optional

foo: Any = None

if isinstance(foo, int | None):
    reveal_type(foo)

if isinstance(foo, Union[int, None]):
    reveal_type(foo)

if isinstance(foo, Optional[int]):
    reveal_type(foo)

bar = int | None
baz = Optional[int]
if isinstance(foo, bar):
    reveal_type(foo)

if isinstance(foo, baz):
    reveal_type(foo)
```

Output for the above with mypy v1.11.1:
```
src/union.py:6: note: Revealed type is "Union[builtins.int, None]"
src/union.py:8: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
src/union.py:9: note: Revealed type is "Any"
src/union.py:11: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
src/union.py:12: note: Revealed type is "Any"
src/union.py:17: note: Revealed type is "Union[builtins.int, None]"
src/union.py:19: error: Parameterized generics cannot be used with class or instance checks  [misc]
src/union.py:19: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
src/union.py:20: note: Revealed type is "Any"
Found 4 errors in 1 file (checked 1 source file)
```

Running the above with pyright:
```
.../src/union.py
  .../src/union.py:6:17 - information: Type of "foo" is "int | None"
  .../src/union.py:9:17 - information: Type of "foo" is "int | None"
  .../src/union.py:12:17 - information: Type of "foo" is "int | None"
  .../src/union.py:17:17 - information: Type of "foo" is "int | None"
  .../src/union.py:20:17 - information: Type of "foo" is "int | None"
0 errors, 0 warnings, 5 informations
```

Summary:
- when using the PEP 604-syntax the type is correctly narrowed to `Union[builtins.int, None]`
- when using the `typing.Union` and/or `typing.Optional` syntax the type is not narrowed and remains `Any` (and errors are shown)

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -582,7 +582,10 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
                     and not node.node.no_args
                     and not (
                         isinstance(union_target := get_proper_type(node.node.target), UnionType)
-                        and union_target.uses_pep604_syntax
+                        and (
+                            union_target.uses_pep604_syntax
+                            or self.chk.options.python_version >= (3, 10)
+                        )
                     )
                 ):
                     self.msg.type_arguments_not_allowed(e)

--- a/test-data/unit/check-unions.test
+++ b/test-data/unit/check-unions.test
@@ -1345,3 +1345,30 @@ x: Union[C1, C2, C3, C4, C5, C6, C7, C8, C9, C10, C11]
 y: Union[C1, C2, C3, C4, C5, C6, C7, C8, C9, C10, C11, None]
 x = y  # E: Incompatible types in assignment (expression has type "Union[C1, C2, C3, C4, C5, <6 more items>, None]", variable has type "Union[C1, C2, C3, C4, C5, <6 more items>]") \
        # N: Item in the first union not in the second: "None"
+
+[case testTypeAliasWithOldUnionIsInstance]
+# flags: --python-version 3.10
+from typing import Union
+SimpleAlias = Union[int, str]
+
+def foo(x: Union[int, str, tuple]):
+    # TODO: fix the typeshed stub for isinstance
+    if isinstance(x, SimpleAlias):  # E: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "type"
+        reveal_type(x)  # N: Revealed type is "Union[builtins.int, builtins.str, builtins.tuple[Any, ...]]"
+    else:
+        reveal_type(x)  # N: Revealed type is "Union[builtins.int, builtins.str, builtins.tuple[Any, ...]]"
+[builtins fixtures/tuple.pyi]
+
+
+[case testTypeAliasWithOldUnionIsInstancePython39]
+# flags: --python-version 3.9
+from typing import Union
+SimpleAlias = Union[int, str]
+
+def foo(x: Union[int, str, tuple]):
+    if isinstance(x, SimpleAlias):  # E: Parameterized generics cannot be used with class or instance checks \
+                                    # E: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "type"
+        reveal_type(x)  # N: Revealed type is "Union[builtins.int, builtins.str, builtins.tuple[Any, ...]]"
+    else:
+        reveal_type(x)  # N: Revealed type is "Union[builtins.int, builtins.str, builtins.tuple[Any, ...]]"
+[builtins fixtures/tuple.pyi]

